File: examples/dude/dude.py
#!/usr/bin/env python3
"""dude (CLI driver) — install the dude package on first boot, enable it, read it back.

The Python/CLI sibling of dude.ts. Drives the `quickchr` CLI via subprocess: the
library shows `installPackage()` (post-boot); the CLI installs at first boot via
`--add-package` (there is no post-boot "install on a running machine" CLI command
yet — see the example README's "friction found"). Works on x86 and arm64.

Run:  uv run examples/dude/dude.py [--channel stable] [--no-cleanup]
      (uv is preferred over a venv; this script is stdlib-only.)
"""

import argparse
import os
import random
import subprocess
import sys


def example_name(slug: str) -> str:
    return f"examples-{slug}-{os.getpid():x}{random.randint(0, 0xFFFF):x}"


def main() -> int:
    ap = argparse.ArgumentParser(description="install + enable the dude package on a CHR")
    ap.add_argument("--channel", default="stable")
    ap.add_argument("--no-cleanup", action="store_true", help="leave the CHR running")
    ap.add_argument(
        "--quickchr",
        help='override the quickchr command, e.g. "bun run ../../src/cli/index.ts --"',
    )
    args = ap.parse_args()

    # Resolve quickchr: explicit override, else env, else the installed binary.
    if args.quickchr:
        qc = args.quickchr.split()
    elif os.environ.get("QUICKCHR"):
        qc = os.environ["QUICKCHR"].split()
    else:
        qc = ["quickchr"]

    name = example_name("dude")

    def run(*cli: str) -> str:
        out = subprocess.run([*qc, *cli], capture_output=True, text=True)
        if out.returncode != 0:
            sys.stderr.write(out.stderr)
            raise SystemExit(f"quickchr {cli[0]} failed (exit {out.returncode})")
        return out.stdout

    print(f"→ starting {name} with --add-package dude (downloads + reboots once)…")
    run("start", name, "--channel", args.channel, "--no-secure-login",
        "--add-package", "dude", "--mem", "256")
    try:
        run("exec", name, "/dude/set enabled=yes")
        print(run("exec", name, "/dude/print"))
    finally:
        if not args.no_cleanup:
            run("remove", name)
        else:
            print(f"--no-cleanup: '{name}' left running")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/dude/dude.ts
#!/usr/bin/env bun
/**
 * dude — ground an optional RouterOS *package* and its config against a real router
 *
 * A richer grounding loop than `grounding`: install the `dude` server package
 * (`installPackage` downloads the arch-correct `.npk` from MikroTik and reboots
 * to activate), then enable The Dude server and read the setting back. Shows that
 * a package-gated subsystem (`/dude`) only appears once its package is present.
 *
 * Works on **both x86 and arm64** — MikroTik ships `dude-<ver>-arm64.npk`
 * alongside the x86 build (verified 7.21.1–7.23.1), and quickchr's package
 * resolver picks the right one for the host-native arch (omitted here = auto).
 *
 * Run:  bun run examples/dude/dude.ts
 * Time: ~50–90 s (install downloads the package and reboots once).
 */
import { QuickCHR } from "../../src/index.ts";
import { check, exampleMachineName, runExample } from "../lib.ts";

if (import.meta.main) {
	await runExample(async (track) => {
		const chr = track(
			await QuickCHR.start({
				name: exampleMachineName("dude"),
				channel: "stable",
				secureLogin: false,
				mem: 256,
			}),
		);

		// installPackage downloads the arch-correct package, uploads it, and reboots.
		// Empty return = the package wasn't found for this version/arch → re-ground,
		// don't silently skip.
		const installed = await chr.installPackage("dude");
		check(installed.includes("dude"), "dude package did not install");

		// The /dude menu only exists once the package is present.
		const before = (await chr.rest("/dude")) as { enabled?: string };
		check(typeof before === "object" && before !== null, "/dude should exist after install");

		// Config round-trip: enable the server, read it back.
		await chr.exec("/dude/set enabled=yes");
		const after = (await chr.rest("/dude")) as { enabled?: string };
		check(after.enabled === "true", `expected /dude enabled=true, got ${after.enabled}`);

		console.log(`  dude installed and enabled (enabled=${after.enabled})`);
	});
}
